### ISSUE TITLE
njs module with boringssl error

### ISSUE THREAD

**Issue by vndroid** [nginx/njs#499]
### build boringssl

```ash
apk add --no-cache libunwind-dev build-base cmake perl git linux-headers pkgconfig
git clone https://boringssl.googlesource.com/boringssl \
mkdir boringssl/build \
cd boringssl/build \
cmake .. \
make
```

the part refers to [Our Roadmap for QUIC and HTTP/3 Support in NGINX](https://www.nginx.com/blog/our-roadmap-quic-http-3-support-nginx/) 

### build njs standalone

```ash
apk add --no-cache git mercurial perl-dev pcre-dev zlib-dev libxslt-dev gd-dev openssl-dev geoip-dev readline-dev expect
mkdir /src && cd /src
hg clone https://hg.nginx.org/njs
cd /njs
./configure \
--cc-opt="-I/src/boringssl/include" \
--ld-opt="-L/src/boringssl/build/ssl -L/src/boringssl/build/crypto"
make
```

the error 

```
external/njs_webcrypto_module.c: In function 'njs_ext_derive':
external/njs_webcrypto_module.c:1548:20: error: implicit declaration of function 'EVP_PKEY_new_mac_key'; did you mean 'EVP_PKEY_CTX_get0_pkey'? [-Werror=implicit-function-declaration]
 1548 |             pkey = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, k, length);
      |                    ^~~~~~~~~~~~~~~~~~~~
      |                    EVP_PKEY_CTX_get0_pkey
external/njs_webcrypto_module.c:1548:41: error: 'EVP_PKEY_HMAC' undeclared (first use in this function); did you mean 'EVP_PKEY_RSA2'?
 1548 |             pkey = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, k, length);
      |                                         ^~~~~~~~~~~~~
      |                                         EVP_PKEY_RSA2
external/njs_webcrypto_module.c:1548:41: note: each undeclared identifier is reported only once for each function it appears in
external/njs_webcrypto_module.c: In function 'njs_ext_import_key':
external/njs_webcrypto_module.c:1859:37: error: 'EVP_PKEY_HMAC' undeclared (first use in this function); did you mean 'EVP_PKEY_RSA2'?
 1859 |         pkey = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, key_data.start,
      |                                     ^~~~~~~~~~~~~
      |                                     EVP_PKEY_RSA2
cc1: all warnings being treated as errors
make: *** [build/Makefile:768: build/external/njs_webcrypto_module.o] Error 1
make: *** Waiting for unfinished jobs....
```

### build nginx with njs as dynamic module

```ash
cd /src
hg clone https://hg.nginx.org/nginx-quic
cd nginx-quic
hg update quic
auto/configure --prefix=/etc/nginx \
    --sbin-path=/usr/sbin/nginx \
    --modules-path=/usr/lib/nginx/modules \
    --conf-path=/etc/nginx/nginx.conf \
    --error-log-path=/var/log/nginx/error.log \
    --http-log-path=/var/log/nginx/access.log \
    --pid-path=/var/run/nginx.pid \
    --lock-path=/var/run/nginx.lock \
    --http-client-body-temp-path=/var/cache/nginx/client_temp \
    --http-proxy-temp-path=/var/cache/nginx/proxy_temp \
    --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp \
    --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp \
    --http-scgi-temp-path=/var/cache/nginx/scgi_temp \
    --with-perl_modules_path=/usr/lib/perl5/vendor_perl \
    --user=nginx \
    --group=nginx \
    --build=nginx-quic \
    --with-compat \
    --with-file-aio \
    --with-threads \
    --with-http_addition_module \
    --with-http_auth_request_module \
    --with-http_dav_module \
    --with-http_flv_module \
    --with-http_gunzip_module \
    --with-http_gzip_static_module \
    --with-http_mp4_module \
    --with-http_random_index_module \
    --with-http_realip_module \
    --with-http_secure_link_module \
    --with-http_slice_module \
    --with-http_ssl_module \
    --with-http_stub_status_module \
    --with-http_sub_module \
    --with-http_v2_module \
    --with-http_v3_module \
    --with-mail \
    --with-mail_ssl_module \
    --with-stream \
    --with-stream_realip_module \
    --with-stream_ssl_module \
    --with-stream_ssl_preread_module \
    --with-stream_quic_module \
    --with-http_xslt_module=dynamic \
    --with-http_image_filter_module=dynamic \
    --with-http_geoip_module=dynamic \
    --with-http_perl_module=dynamic \
    --with-stream_geoip_module=dynamic \
    --add-dynamic-module=/src/njs/nginx \
    --with-cc-opt='-I/src/boringssl/include -Os -fomit-frame-pointer -g' \
    --with-ld-opt='-L/src/boringssl/build/ssl -L/src/boringssl/build/crypto -Wl,--as-needed,-O1,--sort-common'
make -j$(getconf _NPROCESSORS_ONLN)
```

the error

```
	-o objs/addon/external/njs_webcrypto_module.o \
	/src/njs/nginx/../external/njs_webcrypto_module.c
/src/njs/nginx/../external/njs_webcrypto_module.c: In function 'njs_ext_derive':
/src/njs/nginx/../external/njs_webcrypto_module.c:1548:20: error: implicit declaration of function 'EVP_PKEY_new_mac_key'; did you mean 'EVP_PKEY_CTX_get0_pkey'? [-Werror=implicit-function-declaration]
 1548 |             pkey = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, k, length);
      |                    ^~~~~~~~~~~~~~~~~~~~
      |                    EVP_PKEY_CTX_get0_pkey
/src/njs/nginx/../external/njs_webcrypto_module.c:1548:41: error: 'EVP_PKEY_HMAC' undeclared (first use in this function); did you mean 'EVP_PKEY_RSA2'?
 1548 |             pkey = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, k, length);
      |                                         ^~~~~~~~~~~~~
      |                                         EVP_PKEY_RSA2
/src/njs/nginx/../external/njs_webcrypto_module.c:1548:41: note: each undeclared identifier is reported only once for each function it appears in
/src/njs/nginx/../external/njs_webcrypto_module.c: In function 'njs_ext_import_key':
/src/njs/nginx/../external/njs_webcrypto_module.c:1859:37: error: 'EVP_PKEY_HMAC' undeclared (first use in this function); did you mean 'EVP_PKEY_RSA2'?
 1859 |         pkey = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, key_data.start,
      |                                     ^~~~~~~~~~~~~
      |                                     EVP_PKEY_RSA2
cc1: all warnings being treated as errors
make[1]: *** [objs/Makefile:2291: objs/addon/external/njs_webcrypto_module.o] Error 1
make[1]: Leaving directory '/src/nginx-quic'
make: *** [Makefile:10: build] Error 2
```

Please help me to solve it.

**Comment by xeioex** [comment]
It is a duplicate of https://github.com/nginx/njs/issues/455, but thanks for the reproduction steps.

**Comment by xeioex** [comment]
Hi @vndroid 

Feel free to test the following [patch](https://gist.github.com/xeioex/e1aa6b83e8ff2b89092f6102bc13f191)

**Comment by vndroid** [comment]
@xeioex Will this feature be supported in the mainline in the future?

**Comment by xeioex** [comment]
Hi @vndroid,

Yes, it will be.

**Comment by vndroid** [comment]
> Hi @vndroid,
> 
> Yes, it will be.

Hi @xeioex ,

After compiling with the patched version, there is an error message when configuring test.

```
/ # nginx -t
2022/05/18 06:58:53 [emerg] 10#10: dlopen() "/etc/nginx/modules/ngx_http_js_module.so" failed (Error relocating /etc/nginx/modules/ngx_http_js_module.so: EVP_PKEY_CTX_set_signature_md: symbol not found) in /etc/nginx/nginx.conf:11
nginx: [emerg] dlopen() "/etc/nginx/modules/ngx_http_js_module.so" failed (Error relocating /etc/nginx/modules/ngx_http_js_module.so: EVP_PKEY_CTX_set_signature_md: symbol not found) in /etc/nginx/nginx.conf:11
nginx: configuration file /etc/nginx/nginx.conf test failed
/ # ls /etc/nginx/modules/ngx_http_js_module.so
```

Environment

- Alpine:3.15
- BoringSSL
- NGiNX-quic latest
- njs latest

Why is that?

**Comment by xeioex** [comment]
@vndroid  could you please provide the complete nginx configure log?

**Comment by xeioex** [comment]
@vndroid 

1) by default boringssl is compiled as a static library
2) during nginx compilation the library is linked against nginx binary
3) nginx dynamic modules use symbols from the nginx binary, but a dynamic module cannot use static symbols from the binary

I see several options:
1) [compile](https://boringssl.googlesource.com/boringssl/+/HEAD/BUILDING.md) boringssl as shared library (using `-DBUILD_SHARED_LIBS=1`). Ensure that boringssl library path is available during nginx binary start (for example using  LD_LIBRARY_PATH variable).
2) compile njs module as a built-in module (`--add-module`), here you may use static boringssl

**Comment by xeioex** [comment]
@vndroid 

>remove the lines about js_module, then it works.

the reason for that is simple: only njs module `libcrypto` symbols from boringssl.

**Comment by vndroid** [comment]
Ok, I see. Thank you very much.

**Comment by vndroid** [comment]
@xeioex Is there any way to hide the compiled parameters?

such as

```
    --with-cc-opt='-I/src/boringssl/include -Os -fomit-frame-pointer -g' \
    --with-ld-opt='-L/src/boringssl/build/ssl -L/src/boringssl/build/crypto -Wl,--as-needed,-O1,--sort-common'
```

the parameters `-I/src/boringssl/include` and `-L/src/boringssl/build/ssl -L/src/boringssl/build/crypto` In the final container, these files no longer exist. Is there a way to hide or remove these parameters?
